pkg/oc/cli/cmd/newapp.go

```diff
@@ -148,6 +149,33 @@ type NewAppOptions struct {
 	*ObjectGeneratorOptions
 }
 
+func checkResources(discoveryClient discovery.DiscoveryInterface) error {
+	resources, err := discoveryClient.ServerResources()
+	if err != nil {
+		return fmt.Errorf("Unable to to get list of available resources: %v", err)
+	}
+
+	resourceMap := make(map[string]bool)
+	for _, resourceList := range resources {
+		for _, resource := range resourceList.APIResources {
+			resourceMap[resource.Kind] = true
+		}
+	}
+
+	requiredResources := []string{"BuildConfig", "ImageStream"}
+	missingResources := make([]string, 0)
+	for _, r := range requiredResources {
+		if present := resourceMap[r]; !present {
+			missingResources = append(missingResources, r)
+		}
+	}
+
+	if len(missingResources) > 0 {
+		return fmt.Errorf("Missing required resources: %v", strings.Join(missingResources, `, `))
```

Comment:
"API server is missing required resources: "

Reply:
maybe. If new-app generated a buildconfig, we definitely want this to fail, we don't want to create some, but not all, of the resources.  Creating a DC when the BC that generates the image the DC deploys, doesn't exist, would be bad.

But if you're actually just using new-app to create a DC (`oc new-app centos/mysql-56-centos7`, or `oc new-app -f sometemplatewithoutaBC.yaml`) I suppose it's fair to allow that to proceed.

So I think the ideal solution would iterate the objects we are going to attempt to create, ensure those types all exist in the api server, and if not, abort indicating which resources that new-app wanted to create, were not available in the api server (abort before creating *any* resources).

Reply:
Yup, that last sentence is exactly what I'd expect to happen. The current approach fails always, but like Ben said `oc new-app centos/mysql-56-centos7` should work.